File: packages/core/src/vat.ts
// ─── VAT Module ──────────────────────────────────────────────────────────────
// Pure-function VAT calculation engine for Nigerian VAT per NTA 2025.
// Zero dependencies, deterministic output, banker's rounding on all monetary values.

import type { CalculationExplanation, TaxCategory, VatCalculateOptions, VatResult } from './types';
import { InvalidCategoryError } from './errors';
import { assertNonNegativeFinite, bankersRound, parseIsoDateParts } from './utils';
import { get } from './rates';
import { collectRateSources } from './explain';

// ─── Internal Helpers ────────────────────────────────────────────────────────

/** All recognised VAT categories. */
const ZERO_RATED_CATEGORIES: ReadonlySet<string> = new Set([
  'basic-food',
  'medicine',
  'medical-equipment',
  'medical-services',
  'educational-books',
  'tuition',
  'electricity',
  'export-services',
  'humanitarian-goods',
]);

const EXEMPT_CATEGORIES: ReadonlySet<string> = new Set([
  'residential-rent',
  'public-transport',
  'financial-services',
  'insurance',
]);

const ALL_CATEGORIES: TaxCategory[] = [
  'standard',
  ...Array.from(ZERO_RATED_CATEGORIES) as TaxCategory[],
  ...Array.from(EXEMPT_CATEGORIES) as TaxCategory[],
];

/**
 * Determine the rate type for a given category.
 */
function classifyCategory(category: TaxCategory): 'standard' | 'zero-rated' | 'exempt' {
  if (ZERO_RATED_CATEGORIES.has(category)) return 'zero-rated';
  if (EXEMPT_CATEGORIES.has(category)) return 'exempt';
  return 'standard';
}

/**
 * Resolve the legal basis string for a category from the rate data.
 */
function getLegalBasis(category: TaxCategory): string {
  if (category === 'standard') {
    return get('vat.standard.legalBasis') as string;
  }
  if (ZERO_RATED_CATEGORIES.has(category)) {
    return get(`vat.zeroRated.${category}.legalBasis`) as string;
  }
  return get(`vat.exempt.${category}.legalBasis`) as string;
}

function getRateKey(category: TaxCategory): string {
  if (category === 'standard') {
    return 'vat.standard.rate';
  }
  if (ZERO_RATED_CATEGORIES.has(category)) {
    return `vat.zeroRated.${category}.rate`;
  }
  return `vat.exempt.${category}.rate`;
}

/**
 * Validate inputs common to all VAT operations.
 */
function validateInputs(amount: number, category: TaxCategory): void {
  assertNonNegativeFinite('amount', amount);
  if (!ALL_CATEGORIES.includes(category)) {
    throw new InvalidCategoryError(
      `Unknown VAT category "${category}"`,
      ALL_CATEGORIES,
    );
  }
}

// ─── Public API ──────────────────────────────────────────────────────────────

/**
 * Calculate VAT on an amount.
 *
 * - Exclusive (default): net = amount, vat = net × rate, gross = net + vat
 * - Inclusive: gross = amount, net = gross / (1 + rate), vat = gross − net
 *
 * All monetary values are banker's-rounded to 2dp.
 */
export function calculate(options: VatCalculateOptions): VatResult {
  const {
    amount,
    inclusive = false,
    category = 'standard',
    date,
  } = options;

  validateInputs(amount, category);

  const rateType = classifyCategory(category);
  const rate = resolveRate(category, date);
  const legalBasis = getLegalBasis(category);
  const inputVatRecoverable = rateType !== 'exempt';

  let net: number;
  let vat: number;
  let gross: number;

  if (rateType === 'zero-rated' || rateType === 'exempt') {
    // No VAT regardless of inclusive flag
    net = bankersRound(amount);
    vat = 0;
    gross = net;
  } else if (inclusive) {
    // Extract VAT from gross
    gross = bankersRound(amount);
    net = bankersRound(amount / (1 + rate));
    vat = bankersRound(gross - net);
  } else {
    // Add VAT to net
    net = bankersRound(amount);
    vat = bankersRound(net * rate);
    gross = bankersRound(net + vat);
  }

  return {
    net,
    vat,
    gross,
    rate,
    rateType,
    category,
    legalBasis,
    inputVatRecoverable,
  };
}

/**
 * Calculate VAT and return the source-backed reasoning used for the result.
 */
export function explainCalculate(options: VatCalculateOptions): CalculationExplanation<VatResult> {
  const {
    amount,
    inclusive = false,
    category = 'standard',
    date,
  } = options;
  const result = calculate({ amount, inclusive, category, date });
  const rateKeys = [getRateKey(category)];
  const { sources, warnings } = collectRateSources(rateKeys);

  const formula = result.rateType === 'standard'
    ? inclusive
      ? [
          'Gross amount is VAT-inclusive.',
          'Net amount = gross amount / (1 + rate).',
          'VAT = gross amount - net amount.',
        ]
      : [
          'Net amount is VAT-exclusive.',
          'VAT = net amount * rate.',
          'Gross amount = net amount + VAT.',
        ]
    : [
        `VAT = 0 because category "${category}" is ${result.rateType}.`,
        'Gross amount equals net amount.',
      ];

  const assumptions = [
    'Amounts are denominated in Nigerian naira.',
    "Monetary values use banker's rounding to 2 decimal places.",
  ];

  if (result.rateType === 'zero-rated') {
    assumptions.push('The category is zero-rated, so output VAT is 0 and input VAT recoverable remains true.');
  }
  if (result.rateType === 'exempt') {
    assumptions.push('The category is exempt, so output VAT is 0 and input VAT recoverable is false.');
  }
  if (date) {
    assumptions.push(`The requested date ${date} is used for rate-regime selection.`);
  }

  return {
    inputs: { amount, inclusive, category, date },
    result,
    formula,
    assumptions,
    rateKeys,
    sources,
    warnings,
  };
}

/**
 * Extract VAT from a VAT-inclusive amount.
 * Convenience alias for calculate() with inclusive=true.
 */
export function extract(options: VatCalculateOptions): VatResult {
  return calculate({ ...options, inclusive: true });
}

/**
 * Returns true if the category is subject to VAT (standard rate > 0).
 */
export function isTaxable(category: TaxCategory): boolean {
  return classifyCategory(category) === 'standard';
}

/**
 * Returns true if the category is zero-rated (0% VAT, input VAT recoverable).
 */
export function isZeroRated(category: TaxCategory): boolean {
  return classifyCategory(category) === 'zero-rated';
}

/**
 * Returns true if the category is VAT-exempt (0% VAT, input VAT NOT recoverable).
 */
export function isExempt(category: TaxCategory): boolean {
  return classifyCategory(category) === 'exempt';
}

/**
 * Get the VAT rate for a category, optionally for a specific date.
 * Pre-2026 rates use 7.5% standard; post-2026 uses the bundled rate data.
 */
export function getRate(category: TaxCategory, date?: string): number {
  return resolveRate(category, date);
}

/**
 * List all recognised VAT categories.
 */
export function listCategories(): TaxCategory[] {
  return [...ALL_CATEGORIES];
}

// ─── Rate Resolution ─────────────────────────────────────────────────────────

/**
 * Resolve the numeric VAT rate for a category, with optional date-based regime selection.
 * Pre-2026: standard rate is 7.5% (same as post-2026 in current data).
 * Post-2026: uses bundled rate data.
 */
function resolveRate(category: TaxCategory, date?: string): number {
  const rateType = classifyCategory(category);

  if (rateType === 'zero-rated' || rateType === 'exempt') {
    return 0;
  }

  // Date-based regime selection
  if (date) {
    const { year } = parseIsoDateParts(date, 'date');
    if (year < 2026) {
      // Pre-2026: VAT was 7.5% (same rate, but supports future divergence)
      return 0.075;
    }
  }

  // Post-2026 or no date: use bundled rate data
  return get('vat.standard.rate') as number;
}

File: packages/python/src/ngtaxkit/vat.py
"""VAT Module — Pure-function VAT calculation engine for Nigerian VAT per NTA 2025."""

from __future__ import annotations

from typing import cast

from .errors import InvalidCategoryError
from .explain import collect_rate_sources
from .rates import get_float, get_str
from .types import CalculationExplanation, RateType, TaxCategory, VatResult
from .utils import assert_non_negative_finite, bankers_round, parse_iso_date

# ─── Internal Helpers ─────────────────────────────────────────────────────────

ZERO_RATED_CATEGORIES: frozenset[str] = frozenset([
    "basic-food",
    "medicine",
    "medical-equipment",
    "medical-services",
    "educational-books",
    "tuition",
    "electricity",
    "export-services",
    "humanitarian-goods",
])

EXEMPT_CATEGORIES: frozenset[str] = frozenset([
    "residential-rent",
    "public-transport",
    "financial-services",
    "insurance",
])

ALL_CATEGORIES: list[str] = [
    "standard",
    *sorted(ZERO_RATED_CATEGORIES),
    *sorted(EXEMPT_CATEGORIES),
]


def _classify_category(category: str) -> RateType:
    """Determine the rate type for a given category."""
    if category in ZERO_RATED_CATEGORIES:
        return "zero-rated"
    if category in EXEMPT_CATEGORIES:
        return "exempt"
    return "standard"


def _get_legal_basis(category: str) -> str:
    """Resolve the legal basis string for a category from the rate data."""
    if category == "standard":
        return get_str("vat.standard.legalBasis")
    if category in ZERO_RATED_CATEGORIES:
        return get_str(f"vat.zeroRated.{category}.legalBasis")
    return get_str(f"vat.exempt.{category}.legalBasis")


def _get_rate_key(category: str) -> str:
    if category == "standard":
        return "vat.standard.rate"
    if category in ZERO_RATED_CATEGORIES:
        return f"vat.zeroRated.{category}.rate"
    return f"vat.exempt.{category}.rate"


def _validate_inputs(amount: float, category: str) -> None:
    assert_non_negative_finite("amount", amount)
    if category not in ALL_CATEGORIES:
        raise InvalidCategoryError(
            f'Unknown VAT category "{category}"',
            ALL_CATEGORIES,
        )


def _resolve_rate(category: str, date: str | None = None) -> float:
    """Resolve the numeric VAT rate for a category."""
    rate_type = _classify_category(category)
    if rate_type in ("zero-rated", "exempt"):
        return 0.0

    if date:
        year = parse_iso_date(date, "date").year
        if year < 2026:
            return 0.075

    return get_float("vat.standard.rate")


# ─── Public API ───────────────────────────────────────────────────────────────


def calculate(
    amount: float,
    inclusive: bool = False,
    category: str = "standard",
    date: str | None = None,
) -> VatResult:
    """Calculate VAT on an amount.

    - Exclusive (default): net = amount, vat = net × rate, gross = net + vat
    - Inclusive: gross = amount, net = gross / (1 + rate), vat = gross − net
    """
    _validate_inputs(amount, category)
    typed_category = cast(TaxCategory, category)

    rate_type = _classify_category(category)
    rate = _resolve_rate(category, date)
    legal_basis = _get_legal_basis(category)
    input_vat_recoverable = rate_type != "exempt"

    if rate_type in ("zero-rated", "exempt"):
        net = bankers_round(amount)
        vat = 0.0
        gross = net
    elif inclusive:
        gross = bankers_round(amount)
        net = bankers_round(amount / (1 + rate))
        vat = bankers_round(gross - net)
    else:
        net = bankers_round(amount)
        vat = bankers_round(net * rate)
        gross = bankers_round(net + vat)

    return VatResult(
        net=net,
        vat=vat,
        gross=gross,
        rate=rate,
        rate_type=rate_type,
        category=typed_category,
        legal_basis=legal_basis,
        input_vat_recoverable=input_vat_recoverable,
    )


def explain_calculate(
    amount: float,
    inclusive: bool = False,
    category: str = "standard",
    date: str | None = None,
) -> CalculationExplanation:
    """Calculate VAT and return the source-backed reasoning used for the result."""
    result = calculate(amount=amount, inclusive=inclusive, category=category, date=date)
    rate_keys = [_get_rate_key(category)]
    sources, warnings = collect_rate_sources(rate_keys)

    if result["rate_type"] == "standard":
        formula = (
            [
                "Gross amount is VAT-inclusive.",
                "Net amount = gross amount / (1 + rate).",
                "VAT = gross amount - net amount.",
            ]
            if inclusive
            else [
                "Net amount is VAT-exclusive.",
                "VAT = net amount * rate.",
                "Gross amount = net amount + VAT.",
            ]
        )
    else:
        formula = [
            f'VAT = 0 because category "{category}" is {result["rate_type"]}.',
            "Gross amount equals net amount.",
        ]

    assumptions = [
        "Amounts are denominated in Nigerian naira.",
        "Monetary values use banker's rounding to 2 decimal places.",
    ]
    if result["rate_type"] == "zero-rated":
        assumptions.append(
            "The category is zero-rated, so output VAT is 0 and input VAT recoverable remains true."
        )
    if result["rate_type"] == "exempt":
        assumptions.append(
            "The category is exempt, so output VAT is 0 and input VAT recoverable is false."
        )
    if date:
        assumptions.append(f"The requested date {date} is used for rate-regime selection.")

    return CalculationExplanation(
        inputs={
            "amount": amount,
            "inclusive": inclusive,
            "category": category,
            "date": date,
        },
        result=result,
        formula=formula,
        assumptions=assumptions,
        rate_keys=rate_keys,
        sources=sources,
        warnings=warnings,
    )


def extract(
    amount: float,
    category: str = "standard",
    date: str | None = None,
) -> VatResult:
    """Extract VAT from a VAT-inclusive amount. Alias for calculate with inclusive=True."""
    return calculate(amount, inclusive=True, category=category, date=date)


def is_taxable(category: str) -> bool:
    """Returns True if the category is subject to VAT (standard rate > 0)."""
    return _classify_category(category) == "standard"


def is_zero_rated(category: str) -> bool:
    """Returns True if the category is zero-rated (0% VAT, input VAT recoverable)."""
    return _classify_category(category) == "zero-rated"


def is_exempt(category: str) -> bool:
    """Returns True if the category is VAT-exempt (0% VAT, input VAT NOT recoverable)."""
    return _classify_category(category) == "exempt"


def get_rate(category: str, date: str | None = None) -> float:
    """Get the VAT rate for a category, optionally for a specific date."""
    return _resolve_rate(category, date)


def list_categories() -> list[str]:
    """List all recognised VAT categories."""
    return list(ALL_CATEGORIES)
